lib/vector-core/src/stream/batcher.rs

```diff
@@ -0,0 +1,540 @@
+    /// bytes stored in this batch, not taking into account overhead from this
+    /// structure itself. Caller is responsible for ensuring that `I` will fit
+    /// inside the allocation limit.
+    ///
+    /// # Panics
+    ///
+    /// This function will panic if the allocation limit will not store at least
+    /// 1 instance of `I`, as measured by `mem::size_of`.
+    fn new(element_limit: usize, allocation_limit: usize) -> Self {
+        assert!(allocation_limit >= mem::size_of::<I>());
+        Self {
+            allocated_bytes: 0,
+            element_limit,
+            allocation_limit,
+            elements: Vec::with_capacity(element_limit),
+        }
+    }
+
+    /// Unconditionally insert an element into the batch
+    ///
+    /// This function is similar to `push` except that the caller does not need
+    /// to call `has_space` prior to calling this and it will never
+    /// panic. Intended to be used only when insertion must not fail.
+    fn with(mut self, value: I) -> Self {
+        self.allocated_bytes += value.size_of();
+        self.elements.push(value);
+        self
+    }
+
+    /// Decompose the batch
+    ///
+    /// Called by the user when they want to get at the internal store of
+    /// items. Returns a tuple, the first element being the allocated size of
+    /// stored items and the second the store of items.
+    fn destruct(self) -> (usize, Vec<I>) {
+        (self.allocated_bytes, self.elements)
+    }
+
+    /// Whether the batch has space for a new item
+    ///
+    /// This function returns true of there is space both in terms of item count
+    /// and byte count for the given item, false otherwise.
+    fn has_space(&self, value: &I) -> bool {
+        let too_many_elements = self.elements.len() + 1 > self.element_limit;
+        let too_many_bytes = self.allocated_bytes + value.size_of() > self.allocation_limit;
+        !(too_many_elements || too_many_bytes)
+    }
+
+    /// Push an element into the batch
+    ///
+    /// This function pushes an element into the batch. Callers must be sure to
+    /// call `has_space` prior to calling this function and receive a positive
+    /// result.
+    ///
+    /// # Panics
+    ///
+    /// This function will panic if there is not sufficient space in the batch
+    /// for a new element to be inserted.
+    fn push(&mut self, value: I) {
+        assert!(self.has_space(&value));
+        self.allocated_bytes += value.size_of();
+        self.elements.push(value);
+    }
+}
+
+#[pin_project]
+pub struct Batcher<St, Prt, T>
+where
+    Prt: Partitioner,
+{
+    /// The timer to maintain periodic flushes from this `Batcher`
+    timer: T,
+    /// The maximum number of items that are allowed per-batch
+    batch_item_limit: usize,
+    /// The total number of bytes a single batch in this struct is allowed to
+    /// hold. If None effectively unlimited.
+    batch_allocation_limit: usize,
```

Comment:
Is this comment out of date? This member is not an `Option`.

Reply:
It is.

Reply:
Resolved in 762173f43